capsules/extra/src/hc_sr04.rs

```diff
@@ -0,0 +1,158 @@
+/// Syscall driver number.
+use capsules_core::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Distance as usize;
+
+#[derive(Copy, Clone, PartialEq)]
+/// Status of the sensor.
+pub enum Status {
+    Idle,         // Sensor is idle.
+    TriggerPulse, // Sending ultrasonic pulse.
+    EchoStart,    // Interrupt on the rising edge.
+    EchoEnd,      // Interrupt on the falling edge.
+}
+
+/// HC-SR04 Ultrasonic Distance Sensor Driver
+pub struct HcSr04<'a, A: Alarm<'a>> {
+    trig: &'a dyn kernel::hil::gpio::Pin,
+    echo: &'a dyn kernel::hil::gpio::InterruptPin<'a>,
+    alarm: &'a A,
+    start_time: Cell<u32>,
+    end_time: Cell<u32>,
+    status: Cell<Status>,
+    distance_client: OptionalCell<&'a dyn sensors::DistanceClient>,
+}
+
+impl<'a, A: Alarm<'a>> HcSr04<'a, A> {
+    /// Create a new HC-SR04 driver.
+    pub fn new(
+        trig: &'a dyn kernel::hil::gpio::Pin,
+        echo: &'a dyn kernel::hil::gpio::InterruptPin<'a>,
+        alarm: &'a A,
+    ) -> HcSr04<'a, A> {
+        // Setup and return struct.
+        HcSr04 {
+            trig,
+            echo,
+            alarm,
+            start_time: Cell::new(0),
+            end_time: Cell::new(0),
+            status: Cell::new(Status::Idle),
+            distance_client: OptionalCell::empty(),
+        }
+    }
+}
+
+impl<'a, A: Alarm<'a>> Distance<'a> for HcSr04<'a, A> {
+    /// Set the client for distance measurement results.
+    fn set_client(&self, distance_client: &'a dyn DistanceClient) {
+        self.distance_client.set(distance_client);
+    }
+    /// Start a distance measurement.
+    fn read_distance(&self) -> Result<(), ErrorCode> {
+        if self.status.get() == Status::Idle {
+            self.status.set(Status::TriggerPulse);
+            self.trig.set(); // Send the trigger pulse.
+            self.alarm
+                .set_alarm(self.alarm.now(), self.alarm.ticks_from_ms(10));
+            Ok(())
+        } else {
+            Err(ErrorCode::BUSY)
+        }
+    }
+    /// Get the maximum distance the sensor can measure in mm
+    fn get_maximum_distance(&self) -> u32 {
+        // The maximum distance is determined by the maximum pulse width the sensor can detect.
+        // As specified in the datasheet: https://www.handsontec.com/dataspecs/HC-SR04-Ultrasonic.pdf,
+        // the maximum measurable distance is approximately 4 meters.
+        // Convert this to millimeters.
+        4000
+    }
+    /// Get the minimum distance the sensor can measure in mm.
+    fn get_minimum_distance(&self) -> u32 {
+        // The minimum distance is determined by the minimum pulse width the sensor can detect.
+        // As specified in the datasheet: https://www.handsontec.com/dataspecs/HC-SR04-Ultrasonic.pdf,
+        // the minimum measurable distance is approximately 2 cm.
+        // Convert this to millimeters.
+        20
+    }
+}
+
+impl<'a, A: Alarm<'a>> AlarmClient for HcSr04<'a, A> {
+    /// Handle the alarm event.
+    fn alarm(&self) {
```

Comment:
It might be worth considering timeouts. If you don't receive an EchoStart or EchoEnd interrupt in ~50 ms, something has gone wrong and you can fail the measurement. If you do receive a GPIO interrupt, you can cancel the timer event.